### PR TITLE
(0.94.1) Refactor `SplitExplicitFreeSurface`

### PR DESCRIPTION
**Pull request by simone-silvestri**
refactoring of the `SplitExplicitFreeSurface` to comply with issue #3873 

Additional changes included in this PR not described in #3873:
- the grid halos are extended automatically upon passing a grid with a `Connected` topology (`FullyConnected`, `RightConnected` or `LeftConnected`) so there is no need to extend the constructor for grids (such as distributed grids) that have connected topologies.
- All the `ExplicitFreeSurface` relevant code has been moved in the `explicit_free_surface.jl` file

### CODE REVIEW

**Comment by simone-silvestri** [comment]
The `SplitExplicitFreeSurfaces` module is only temporary here for easier manipulation and refactoring of the code. We can remove it if we decide we don't want the free surface in a module

**Comment by simone-silvestri** [comment]
I think the refactor here is complete. All tests pass. This should be ready to review.

File: src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/split_explicit_timesteppers.jl [hunk 1]
```diff
@@ -0,0 +1,151 @@
+"""
+    struct ForwardBackwardScheme
+
+A timestepping scheme used for substepping in the split-explicit free surface solver.
+    
+The equations are evolved as follows:
+```julia
+ηᵐ⁺¹ = ηᵐ - Δτ (∂x Uᵐ + ∂y Vᵐ)
```
Comment (navidcy) [inline src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/split_explicit_timesteppers.jl:8]: is there a `g H` missing here?

Comment (simone-silvestri) [reply src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/split_explicit_timesteppers.jl:8]: The`g H` is in the barotropic velocity equations.

**Review by ali-ramadhan — APPROVED** [review]
Refactor looks super nice and clean to me! Appreciate that it seems easier to account for hydrostatic open BCs now. I just left some minor questions.

File: src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/barotropic_correction.jl [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+end
+
+function barotropic_split_explicit_corrector!(u, v, free_surface, grid)
+    state = free_surface.filtered_state
+    U, V  = free_surface.barotropic_velocities
+    U̅, V̅  = state.U, state.V
+    arch  = architecture(grid)
+
+    # take out "bad" barotropic mode, 
+    # !!!! reusing U̅ and V̅ for U and V have already been restarted
```
Comment (navidcy) [inline src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/barotropic_correction.jl:50]: what does "restarted" mean here?

File: src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/split_explicit_timesteppers.jl [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+```    
+
+where `η̃`, `Ũ` and `Ṽ` are the AB3 time-extrapolated values of free surface, 
+barotropic zonal and meridional velocities, respectively:
+
+```julia
+Ũ = α Uᵐ   + θ Uᵐ⁻¹ + β Uᵐ⁻²
+Ṽ = α Vᵐ   + θ Vᵐ⁻¹ + β Vᵐ⁻²
+η̃ = δ ηᵐ⁺¹ + μ ηᵐ   + γ ηᵐ⁻¹ + ϵ ηᵐ⁻²
+```
```
[the remark this answers is not in the document]
Comment (glwagner) [reply src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/split_explicit_timesteppers.jl]: no we should use ```math.

[the remark this answers is not in the document]
Comment (navidcy) [reply src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/split_explicit_timesteppers.jl]: it was changed... I did ` ```julia` because it wasn't really compiled properly for LaTeX. But @simone-silvestri changed it back. However, eg, `∂x Ũ + ∂y Ṽ` is not LaTeX and it comes out weird. Also we need to break lines with `\\`.

[the remark this answers is not in the document]
Comment (navidcy) [reply src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/split_explicit_timesteppers.jl]: See https://clima.github.io/OceananigansDocumentation/previews/PR3894/appendix/library/#Split-explicit-free-surface-solver

I added the line breaks; should be improved in the next preview.

File: src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/split_explicit_free_surface.jl [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+    SplitExplicitFreeSurface(grid = nothing;
+                             gravitational_acceleration = g_Earth,
+                             substeps = nothing,
+                             cfl = nothing,
+                             fixed_Δt = nothing,
+                             averaging_kernel = averaging_shape_function,
+                             timestepper = ForwardBackwardScheme())
+
+Return a `SplitExplicitFreeSurface` representing an explicit time discretization
+of a free surface dynamics with `gravitational_acceleration`.
```
Comment (glwagner) [inline src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/split_explicit_free_surface.jl:23]: This is a good place to define U, V, and their averaged versions

File: src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/split_explicit_free_surface.jl [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+    FT = eltype(gravitational_acceleration)
+    fractional_step_size, averaging_weights = weights_from_substeps(FT, substeps, averaging_kernel)
+    return FixedSubstepNumber(fractional_step_size, averaging_weights)
+end
+
+# The substeps are calculated dynamically when a cfl without a fixed_Δt is provided
+function split_explicit_substepping(cfl, ::Nothing, ::Nothing, grid, averaging_kernel, gravitational_acceleration)  
+    if isnothing(grid)
+        throw(ArgumentError(string("Need to provide the grid to calculate the barotropic substeps from the cfl. ",
+                                    "For example, SplitExplicitFreeSurface(grid, cfl=0.7, ...)")))
```
Comment (glwagner) [inline src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/split_explicit_free_surface.jl:112]:
```suggestion
        throw(ArgumentError(string("The grid is a required positional argument to SplitExplicitFreeSurface when cfl is specified.",
                                    "For example, SplitExplicitFreeSurface(grid, cfl=0.7, ...)")))
```

However, I think it is better to do this validation in the constructor. Notice that the error really depends on what the constructor looks like. If somebody tries to change the constructor in the future, they may have to change this code as well. Therefore, it belongs in the constructor where the function signature that is required is defined.

File: src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/split_explicit_free_surface.jl [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+    averaging_weights = averaging_weights[1:idx]
+    averaging_weights ./= sum(averaging_weights)
+
+    return Δτ, tuple(averaging_weights...)
+end
+
+Base.summary(s::FixedTimeStepSize)  = string("Barotropic time step equal to $(prettytime(s.Δt_barotropic))")
+Base.summary(s::FixedSubstepNumber) = string("Barotropic fractional step equal to $(s.fractional_step_size) times the baroclinic step")
+
+Base.summary(sefs::SplitExplicitFreeSurface) = string("SplitExplicitFreeSurface with $(summary(sefs.substepping))")
```
Comment (glwagner) [inline src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/split_explicit_free_surface.jl:245]: This summaries are not quite right. The default summary is something like `string(typeof(obj))`. A summary should always try to resemble the code, or how the object is constructed. These summaries are written word; there is no code. For example the summary for `FixedTimeStepSize` _must_ contain the string "FixedTimeStepSize".

File: src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/split_explicit_free_surface.jl [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+
+const ConnectedTopologies = Union{LeftConnected, RightConnected, FullyConnected}
+
+# Extending halos is not allowed with variable time-stepping
+function maybe_extend_halos(TX, TY, grid, ::FixedTimeStepSize) 
+
+    if TX() isa ConnectedTopologies || TY() isa ConnectedTopologies
+        throw(ArgumentError("A variable substepping through a CFL condition is not supported for the `SplitExplicitFreeSurface` on $(summary(grid)). \n
+                             Provide a fixed number of substeps through the `substeps` keyword argument as: \n
+                             `free_surface = SplitExplicitFreeSurface(grid; substeps = N)` where `N::Int`"))
```
Comment (glwagner) [inline src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/split_explicit_free_surface.jl:261]: This is another example of an error that should be in the constructor. Sprinkling the errors / unsupported configurations throughout the code makes it hard to follow the logic of the constructors and user interface.

File: src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/split_explicit_timesteppers.jl [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+
+# The functions `η★` `U★` and `V★` represent the value of free surface, barotropic zonal and meridional velocity at time step m+1/2
+@inline U★(i, j, k, grid, t::ForwardBackwardScheme, Uᵐ) = @inbounds Uᵐ[i, j, k]
+@inline U★(i, j, k, grid, t::AdamsBashforth3Scheme, Uᵐ) = @inbounds t.α * Uᵐ[i, j, k] + t.θ * t.Uᵐ⁻¹[i, j, k] + t.β * t.Uᵐ⁻²[i, j, k]
+
+@inline η★(i, j, k, grid, t::ForwardBackwardScheme, ηᵐ⁺¹) = @inbounds ηᵐ⁺¹[i, j, k]
+@inline η★(i, j, k, grid, t::AdamsBashforth3Scheme, ηᵐ⁺¹) = @inbounds t.δ * ηᵐ⁺¹[i, j, k] + t.μ * t.ηᵐ[i, j, k] + t.γ * t.ηᵐ⁻¹[i, j, k] + t.ϵ * t.ηᵐ⁻²[i, j, k]
+
+@inline advance_previous_velocities!(::ForwardBackwardScheme,   i, j, k, U) = nothing
+@inline advance_previous_free_surface!(::ForwardBackwardScheme, i, j, k, η) = nothing
```
Comment (glwagner) [inline src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/split_explicit_timesteppers.jl:145]: I think other parts of the code call this something like "store_tendencies". That could be changed to "advance_tendencies"? Either way we should try not to create new terminology for existing concepts.

File: src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/step_split_explicit_free_surface.jl [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+# ∂t(U) = - gH∇η + f
+#
+# the free surface field η and its average η̄ are located on `Face`s at the surface (grid.Nz +1). All other intermediate variables
+# (U, V, Ū, V̄) are barotropic fields (`ReducedField`) for which a k index is not defined
+
+@kernel function _split_explicit_free_surface!(grid, Δτ, η, U, V, timestepper)
+    i, j = @index(Global, NTuple)
+    k_top = grid.Nz+1
+    
+    advance_previous_free_surface!(timestepper, i, j, k_top, η)
```
Comment (glwagner) [inline src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/step_split_explicit_free_surface.jl:32]:
```suggestion
    store_previous_free_surface!(timestepper, i, j, k_top, η)
```

File: src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/SplitExplicitFreeSurfaces.jl [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+export SplitExplicitFreeSurface, ForwardBackwardScheme, AdamsBashforth3Scheme
+export FixedSubstepNumber, FixedTimeStepSize
+
+using Oceananigans
+using Oceananigans.Architectures
+using Oceananigans.Fields
+using Oceananigans.Utils
+using Oceananigans.Grids
+using Oceananigans.Operators
```
Comment (glwagner) [inline src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/SplitExplicitFreeSurfaces.jl:11]: All of these files have a huge number of `using` and `import`. Are they all really needed?

File: src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/setup_split_explicit.jl [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    @inbounds Gⱽ[i, j, 1] = Δzᶜᶠᶜ(i, j, 1, grid) * ab2_step_Gv(i, j, 1, grid, Gv⁻, Gvⁿ, χ)
+
+    for k in 2:grid.Nz
+        @inbounds Gᵁ[i, j, 1] += Δzᶠᶜᶜ(i, j, k, grid) * ab2_step_Gu(i, j, k, grid, Gu⁻, Guⁿ, χ)
+        @inbounds Gⱽ[i, j, 1] += Δzᶜᶠᶜ(i, j, k, grid) * ab2_step_Gv(i, j, k, grid, Gv⁻, Gvⁿ, χ)
+    end
+end
+
+@inline ab2_step_Gu(i, j, k, grid, G⁻, Gⁿ, χ::FT) where FT =
+    @inbounds ifelse(peripheral_node(i, j, k, grid, f, c, c), zero(grid), (convert(FT, 1.5) + χ) *  Gⁿ[i, j, k] - G⁻[i, j, k] * (convert(FT, 0.5) + χ))
```
Comment (glwagner) [inline src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/setup_split_explicit.jl:51]: I think these fail when `G⁻` has NaNs, even if `χ = -0.5`. I thought there was another PR that fixed this, but it is either reverted here or that other PR has not been merged (I can't find it...)

Also this function contains 5 or 6 computations on one line. It should be split into multiple lines.

File: src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/split_explicit_timesteppers.jl (hunk 3 above)
[the remark this answers is not in the document]
Comment (simone-silvestri) [reply src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/split_explicit_timesteppers.jl]: > it was changed... I did `` ```julia `` because it wasn't really compiled properly for LaTeX. But @simone-silvestri changed it back. However, eg, `∂x Ũ + ∂y Ṽ` is not LaTeX and it comes out weird. Also we need to break lines with `\\`.

Sorry Navid, I didn't realize, I must have overwritten it by mistake.

File: src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/barotropic_correction.jl (hunk 2 above)
Comment (simone-silvestri) [reply src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/barotropic_correction.jl:50]: I have changed the comment to make it more clear. In practice `U̅` and `V̅` have exhausted their scope because we copied the filtered values in `U` and `V`, to initialize them for the new subcycling step, so `U̅` and `V̅` are used here as "work arrays" to store the vertical integral of u and v

File: src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/setup_split_explicit.jl [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+
+@inline ab2_step_Gv(i, j, k, grid, G⁻, Gⁿ, χ::FT) where FT =
+    @inbounds ifelse(peripheral_node(i, j, k, grid, c, f, c), zero(grid), (convert(FT, 1.5) + χ) *  Gⁿ[i, j, k] - G⁻[i, j, k] * (convert(FT, 0.5) + χ))
+
+# Setting up the RHS for the barotropic step (tendencies of the barotropic velocity components)
+# This function is called after `calculate_tendency` and before `ab2_step_velocities!`
+function setup_free_surface!(model, ::SplitExplicitFreeSurface, χ)
```
[the remark this answers is not in the document]
Comment (simone-silvestri) [reply src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/setup_split_explicit.jl:58]: In light of this comment I have made an additional change. The `setup_free_surface!` is basically just calculating the slow tendencies for `U` and `V`, so it makes sense to unify it to the `compute_free_surface_tendency!` that we already have for the `ExplicitFreeSurface`. However, since we need to do it after the `apply_flux_bc!` (we need to integrate upwards also the BCs) I have moved the `compute_free_surface_tendency!` before where we step the free surface

**Comment by simone-silvestri** [comment]
Thanks for the comments, I should have addressed all points.

File: src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/SplitExplicitFreeSurfaces.jl (hunk 10 above)
Comment (simone-silvestri) [reply src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/SplitExplicitFreeSurfaces.jl:12]: There were, indeed, some spurious `using` statement. I have reduced them to a minimum now

File: src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/initialize_split_explicit_substepping.jl [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+        @inbounds Gᵁ[i, j, 1] += Δzᶠᶜᶜ(i, j, k, grid) * ab2_step_G(i, j, k, grid, locU..., Gu⁻, Guⁿ, χ)
+        @inbounds Gⱽ[i, j, 1] += Δzᶜᶠᶜ(i, j, k, grid) * ab2_step_G(i, j, k, grid, locV..., Gv⁻, Gvⁿ, χ)
+    end
+end
+
+@inline function ab2_step_G(i, j, k, grid, ℓx, ℓy, ℓz, G⁻, Gⁿ, χ::FT) where FT 
+    C₁ = convert(FT, 3/2) + χ
+    C₂ = convert(FT, 1/2) + χ
+    
+    Gⁿ⁺¹ = @inbounds C₁ * Gⁿ[i, j, k] - C₂ * G⁻[i, j, k]
```
Comment (glwagner) [inline src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/initialize_split_explicit_substepping.jl:62]:
```suggestion
    not_euler = C₂ != 0
    Gⁿ⁺¹ = @inbounds C₁ * Gⁿ[i, j, k] - C₂ * G⁻[i, j, k] * not_euler
```

multiplying by `false` eliminates the NaN

(we need this in the other time-stepping code too)

File: src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/split_explicit_free_surface.jl [hunk 14]
```diff
@@ -259,7 +259,7 @@ end
 end
 
 Base.summary(s::FixedTimeStepSize)  = string("FixedTimeStepSize with a barotropic time step equal to $(prettytime(s.Δt_barotropic))")
-Base.summary(s::FixedSubstepNumber) = string("FixedSubstepNumber with $(length(s.averaging_kernel)) substeps")
+Base.summary(s::FixedSubstepNumber) = string("FixedSubstepNumber with $(length(s.averaging_weights)) substeps")
```
Comment (glwagner) [inline src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/split_explicit_free_surface.jl:262]:
```suggestion
Base.summary(s::FixedSubstepNumber) = string("FixedSubstepNumber($(length(s.averaging_weights)))")
```

**Comment by tomchor** [comment]
I think was merged with a bump to 0.94.1 before 0.94 was registered. How do we fix this?

**Comment by glwagner** [comment]
we have to merge 0.94 retroactively. But yeah, this should have been merged just as 0.94, not 0.94.1 @simone-silvestri .

**Comment by navidcy** [comment]
We don't need to fix. It's OK. We move on.

**Comment by simone-silvestri** [comment]
Oh sorry, my bad! I messed up a bit here.
